doc/neps/ufunc-overrides.rst

```diff
@@ -213,83 +370,128 @@ They have indirect interactions, however, because NumPy's
 most numerical classes, the easiest way to override binary operations is
 thus to define ``__array_ufunc__`` and override the corresponding
 Ufunc. The class can then, like :class:`ndarray` itself, define the
-binary operators in terms of Ufuncs. Here, one has to take some care.
-E.g., the simplest implementation would be::
+binary operators in terms of Ufuncs. Here, one has to take some care to
+ensure that one allows for other classes to indicate they are not
+compatible, i.e., implementations should be something like::
 
     class ArrayLike(object):
+        ...
         def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
             ...
             return result
-        ...
-        def __mul__(self, other):
-            return self.__array_ufunc__(np.multiply, '__call__', self, other)
-
-Suppose now, however, that ``other`` is class that does not know how to
-deal with arrays and ufuncs, but does know how to do multiplication::
 
-    class MyObject(object):
-        __array_ufunc__ = None
+        # Option 1: call ufunc directly
```

Comment:
This is really useful (I think this will help me rewrite `Quantity.__array_ufunc__` in a better way ;-), but also leaves me slightly confused: in this gist, @shoyer uses option 2, where we call `__array_ufunc__` inside the binop (but write `__array_ufunc__` such that it returns `NotImplemented` if another function defines `__array_ufunc__` and it cannot handle it). I like this most. However, `ndarray` currently does something much closer to option 1: see https://github.com/numpy/numpy/pull/8247/files#diff-a76a87b213692b2a554a63aebea985caR89

So, which should I pick? Here, we should also consider that we might well end up rewriting the `ndarray` code to use `ndarray.__array_ufunc__`...